airbyte-integrations/connectors/source-stripe/unit_tests/integration/test_events.py

```diff
@@ -0,0 +1,391 @@
+    return SourceStripe(catalog)
+
+
+def _create_builders() -> Tuple[RecordBuilder, HttpResponseBuilder]:
+    return create_builders_from_resource(
+        find_template("events", __file__),
+        FieldPath("data"),
+        record_id_path=FieldPath("id"),
+        record_cursor_path=FieldPath("created"),
+        pagination_strategy=StripePaginationStrategy(),
+    )
+
+
+def _a_record() -> RecordBuilder:
+    return _create_builders()[0]
+
+
+def _a_response() -> HttpResponseBuilder:
+    return _create_builders()[1]
+
+
+def _read(
+    config: ConfigBuilder,
+    sync_mode: SyncMode,
+    state: Optional[Dict[str, Any]] = None,
+    expecting_exception: bool = False
+) -> EntrypointOutput:
+    catalog = _catalog(sync_mode)
+    return read(_source(catalog), config.build(), catalog, state, expecting_exception)
+
+
+@freezegun.freeze_time(_NOW.isoformat())
+class FullRefreshTest(TestCase):
+
+    @HttpMocker()
+    def test_given_one_page_when_read_then_return_record(self, http_mocker: HttpMocker) -> None:
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
+                query_params={"created[gte]": str(int(_A_START_DATE.timestamp())), "created[lte]": str(int(_NOW.timestamp())), "limit": 100},
+                headers=_AUTHENTICATION_HEADERS
+            ),
+            _a_response().with_record(_a_record()).with_record(_a_record()).build()
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE))
+        assert len(output.records) == 2
+
+    @HttpMocker()
+    def test_given_many_pages_when_read_then_return_records(self, http_mocker: HttpMocker) -> None:
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
+                query_params={"created[gte]": str(int(_A_START_DATE.timestamp())), "created[lte]": str(int(_NOW.timestamp())), "limit": 100},
+                headers=_AUTHENTICATION_HEADERS,
+            ),
+            _a_response().with_pagination().with_record(_a_record().with_id("last_record_id_from_first_page")).build()
+        )
+        http_mocker.get(
+            HttpRequest(
+                url="https://api.stripe.com/v1/events",
+                query_params={
+                    "starting_after": "last_record_id_from_first_page",
+                    "created[gte]": str(int(_A_START_DATE.timestamp())),
+                    "created[lte]": str(int(_NOW.timestamp())),
+                    "limit": 100,
+                },
+                headers=_AUTHENTICATION_HEADERS,
+            ),
+            _a_response().with_record(_a_record()).with_record(_a_record()).build()
+        )
+        output = self._read(_config().with_start_date(_A_START_DATE))
+        assert len(output.records) == 3
+
+    @HttpMocker()
+    def test_given_start_date_before_30_days_stripe_limit_and_slice_range_when_read_then_perform_request_before_30_days(self, http_mocker: HttpMocker) -> None:
```

Comment:
For test cases like this that are tied specifically to Stripe, feels like a good idea to link to documentation and possibly provide a comment with additional context (e.g. in case the documentation gets stale). I also wonder if we should organize them differently than tests that cover behavior that we'd expect all sources to follow. I think it would be awesome of we could make the integration test suite a sort of living document that highlights caveats/interesting aspects of a specific source, so that people who are maintaining or debugging the connector can get a sense of what makes this connector different.

Reply:
The method name is definitely not enough. I'll update the test to add URL doc and explanation

Reply:
WDYT about separating tests that feel specific to Stripe from tests that we expect to implement for all sources?

Reply:
Can you give me an example of tests that we expect all sources to implement vs source specific? I can be for another source than Stripe. I just want to see what we can generify. Right now, I'm leaning to "we can't generify anything" as even things I would take for granted like record seclection is not implemented by all the streams (if the API returns a list for example)

Reply:
At a glance here are some that seem more generic to me, but tell me if I'm wrong:
- `test_given_lookback_window_when_read_then_request_before_start_date`
- `test_given_slice_range_when_read_then_perform_multiple_requests`
- `test_given_http_status_400_when_read_then_stream_is_ignored`
- `test_given_http_status_401_when_read_then_stream_is_incomplete` << not totally confident about this one
- `test_given_slice_range_when_read_then_perform_multiple_requests`
- `test_when_read_then_validate_availability_for_full_refresh_and_incremental`
... and lots more. Basically it feels like these could be abstractmethods that we could expect integration tests for each source to override.

Reply:
It feels like a good basis for https://github.com/airbytehq/airbyte/issues/33417. I'll add these to the ticket

Reply:
Thank you!